lib/ddtrace/profiling/pprof/builder.rb

```diff
@@ -0,0 +1,181 @@
+require 'ddtrace/profiling/pprof/message_set'
+require 'ddtrace/profiling/pprof/pprof_pb'
+require 'ddtrace/profiling/pprof/string_table'
+
+module Datadog
+  module Profiling
+    module Pprof
+      # Generic profile building behavior
+      class Builder
+        DESC_FRAME_OMITTED = 'frame omitted'.freeze
+        DESC_FRAMES_OMITTED = 'frames omitted'.freeze
+        VALUE_TYPE_WALL = 'wall'.freeze
+        VALUE_UNIT_NANOSECONDS = 'nanoseconds'.freeze
+
+        attr_reader \
+          :events,
+          :functions,
+          :locations,
+          :sample_types,
+          :samples,
+          :string_table
+
+        def initialize(events)
+          @events = events
+          @profile = nil
+          @sample_types = []
+          @samples = []
+          @mappings = []
+          @locations = MessageSet.new
+          @functions = MessageSet.new
+          @string_table = StringTable.new
+        end
+
+        def to_profile
+          @profile ||= build_profile(@events)
+        end
+
+        def build_profile(events)
+          @sample_types = build_sample_types
+          @samples = group_events(events) do |event, values|
+            build_sample(event, values)
+          end
+          @mappings = build_mappings
+
+          Perftools::Profiles::Profile.new(
+            sample_type: @sample_types,
+            sample: @samples,
+            mapping: @mappings,
+            location: @locations.messages,
+            function: @functions.messages,
+            string_table: @string_table.strings
+          )
```

Comment:
Yeah so I think you build in the way an event type constructs a `Profile`, where you want all your event types in a single `Profile`.